File: python/console_student_app.py
from student_models import GraduateStudent, PhDStudent
from student_validator import StudentValidator


class ConsoleStudentApp:
    """Konsol rejimi üçün OOP əsaslı tətbiq sinifi."""

    def __init__(self):
        self.validator = StudentValidator()

    def read_non_empty_input(self, prompt):
        while True:
            value = input(prompt).strip()
            if value:
                return value
            print("Invalid entry: this field cannot be empty. Please enter a valid value.")

    def read_valid_phone(self, prompt):
        while True:
            phone = input(prompt).strip()
            if not phone:
                print("Invalid entry: phone number cannot be empty.")
                continue
            if self.validator.is_valid_phone(phone):
                return phone
            print("Invalid phone number. Use digits, spaces, dashes, and optional leading +.")

    def read_valid_email(self, prompt):
        while True:
            email = input(prompt).strip()
            if not email:
                print("Invalid entry: email cannot be empty.")
                continue
            if self.validator.is_valid_email(email):
                return email
            print("Invalid email format. Example: user@example.com")

    def run(self):
        print("Graduate student data entry:")
        gs_name = self.read_non_empty_input("Enter the name of the graduate student: ")
        gs_surname = self.read_non_empty_input("Enter the surname of the graduate student: ")
        gs_phone = self.read_valid_phone("Enter the phone number of the graduate student: ")
        gs_email = self.read_valid_email("Enter the email of the graduate student: ")
        gs_university = self.read_non_empty_input("Enter the university of the graduate student: ")

        graduate_student = GraduateStudent(gs_name, gs_surname, gs_phone, gs_email, gs_university)
        print()
        graduate_student.show_info()

        print()
        print("PhD student data entry:")
        phd_name = self.read_non_empty_input("Enter the name of the PhD student: ")
        phd_surname = self.read_non_empty_input("Enter the surname of the PhD student: ")
        phd_phone = self.read_valid_phone("Enter the phone number of the PhD student: ")
        phd_email = self.read_valid_email("Enter the email of the PhD student: ")
        phd_university = self.read_non_empty_input("Enter the university of the PhD student: ")
        phd_research_topic = self.read_non_empty_input("Enter the research topic of the PhD student: ")

        phd_student = PhDStudent(phd_name, phd_surname, phd_phone, phd_email, phd_university, phd_research_topic)
        print()
        phd_student.show_info()


def run_console_mode():
    ConsoleStudentApp().run()

File: demo/src/main/java/com/example/ConsoleStudentApp.java
package com.example;

import java.util.Scanner;

public class ConsoleStudentApp {
    private final StudentValidator validator = new StudentValidator();

    public void run() {
        try (Scanner scanner = new Scanner(System.in)) {
            System.out.println("Graduate student data entry:");
            String gsName = readNonEmptyInput(scanner, "Enter the name of the graduate student: ");
            String gsSurname = readNonEmptyInput(scanner, "Enter the surname of the graduate student: ");
            String gsPhone = readValidPhone(scanner, "Enter the phone number of the graduate student: ");
            String gsEmail = readValidEmail(scanner, "Enter the email of the graduate student: ");
            String gsUniversity = readNonEmptyInput(scanner, "Enter the university of the graduate student: ");

            Student graduateStudent = new GraduateStudent(gsName, gsSurname, gsPhone, gsEmail, gsUniversity);
            System.out.println();
            graduateStudent.showInfo();

            System.out.println();
            System.out.println("PhD student data entry:");
            String phdName = readNonEmptyInput(scanner, "Enter the name of the PhD student: ");
            String phdSurname = readNonEmptyInput(scanner, "Enter the surname of the PhD student: ");
            String phdPhone = readValidPhone(scanner, "Enter the phone number of the PhD student: ");
            String phdEmail = readValidEmail(scanner, "Enter the email of the PhD student: ");
            String phdUniversity = readNonEmptyInput(scanner, "Enter the university of the PhD student: ");
            String phdResearchTopic = readNonEmptyInput(scanner, "Enter the research topic of the PhD student: ");

            Student phdStudent = new PhDStudent(phdName, phdSurname, phdPhone, phdEmail, phdUniversity, phdResearchTopic);
            System.out.println();
            phdStudent.showInfo();
        }
    }

    private String readNonEmptyInput(Scanner scanner, String prompt) {
        while (true) {
            System.out.print(prompt);
            String input = scanner.nextLine().trim();
            if (validator.isValidInput(input)) {
                return input;
            }
            System.out.println("Invalid entry: this field cannot be empty. Please enter a valid value.");
        }
    }

    private String readValidPhone(Scanner scanner, String prompt) {
        while (true) {
            String phone = readNonEmptyInput(scanner, prompt);
            if (validator.isValidPhone(phone)) {
                return phone;
            }
            System.out.println("Invalid phone number. Use digits, spaces, dashes, and optional leading +.");
        }
    }

    private String readValidEmail(Scanner scanner, String prompt) {
        while (true) {
            String email = readNonEmptyInput(scanner, prompt);
            if (validator.isValidEmail(email)) {
                return email;
            }
            System.out.println("Invalid email format. Example: user@example.com");
        }
    }
}
